source-oracle/source_oracle/__init__.py

```diff
@@ -0,0 +1,77 @@
+from logging import Logger
+from typing import Callable, Awaitable
+import oracledb
+
+from estuary_cdk.flow import ConnectorSpec
+from estuary_cdk.capture import (
+    BaseCaptureConnector,
+    Request,
+    Task,
+    common,
+    request,
+    response,
+)
+from estuary_cdk.http import HTTPMixin
+
+from .api import (
+    create_pool,
+    fetch_tables,
+    fetch_columns,
+    fetch_page,
+    fetch_changes,
+)
+
+from .resources import (
+    all_resources,
+    validate_flashback
+)
+from .models import (
+    ConnectorState,
+    EndpointConfig,
+    ResourceConfig,
+)
+
+
+class Connector(
+    BaseCaptureConnector[EndpointConfig, ResourceConfig, ConnectorState],
+    HTTPMixin,
+):
+    def request_class(self):
+        return Request[EndpointConfig, ResourceConfig, ConnectorState]
+
+    async def spec(self, log: Logger, _: request.Spec) -> ConnectorSpec:
+        return ConnectorSpec(
+            documentationUrl="https://go.estuary.dev/source-oracle-batch",
+            configSchema=EndpointConfig.model_json_schema(),
+            oauth2=None,
+            resourceConfigSchema=ResourceConfig.model_json_schema(),
+            resourcePathPointers=ResourceConfig.PATH_POINTERS,
+        )
+
+    async def discover(
+        self, log: Logger, discover: request.Discover[EndpointConfig]
+    ) -> response.Discovered[ResourceConfig]:
+        pool = create_pool(discover.config)
+        resources = await all_resources(log, self, discover.config, pool)
+        return common.discovered(resources)
+
+    async def validate(
+        self,
+        log: Logger,
+        validate: request.Validate[EndpointConfig, ResourceConfig],
+    ) -> response.Validated:
+        pool = create_pool(validate.config)
+        await validate_flashback(log, validate.config, pool)
```

Comment:
Do we always run validate before Open and Discover?

Reply:
I think that we don't need to include it in discovery, but the main open RPC might still want to run this validation as well. That's how we handle prerequisites in the SQL CDC connectors for instance, running the checks on Validate() and Pull() both.

It's kind of a judgement call, the concern is the scenario where the user has a long-running capture which has been ticking along happily, and then they make some dangerous changes to their database settings so that it no longer passes our validation checks. Is it better to keep things running and potentially not surface the new issue to the user at all, or is it better for the capture to immediately (or at least on the next task restart) begin failing so that the user can potentially be alerted? I have no idea what the right answer here is.

Reply:
I think that makes sense to me, will add it to Open as well 👍🏽